test/Test/Database/LSMTree/Common.hs

```diff
@@ -1,61 +0,0 @@
-{-# LANGUAGE ScopedTypeVariables #-}
-{-# LANGUAGE TypeApplications    #-}
-module Test.Database.LSMTree.Common (tests) where
-
-import           Control.DeepSeq (NFData, deepseq)
-import qualified Data.ByteString as BS
-import           Data.Proxy (Proxy (..))
-import           Data.Word (Word64)
-import           Database.LSMTree.Common (SomeSerialisationConstraint (..))
-import           Test.QuickCheck.Instances ()
-import           Test.Tasty
-import           Test.Tasty.QuickCheck
-
-tests :: TestTree
-tests = testGroup "Database.LSMTree.Model.Common"
-    [ testGroup "SomeSerialisationConstraint Word64"
-       -- Note: unfortunately Arbitrary Word64 doesn't generate uniformly from whole range.
-        [ decodeEncodeRoundtrip (Proxy @Word64)
-        , totalDeserialiseProp (Proxy @Word64)
-        , preservesOrderProp (Proxy @Word64)
-        ]
-
-    , testGroup "SomeSerialisationConstraint ByteString"
-        [ decodeEncodeRoundtrip (Proxy @BS.ByteString)
-        , totalDeserialiseProp (Proxy @BS.ByteString)
-        , preservesOrderProp (Proxy @BS.ByteString)
-        ]
-    ]
-
--- | decode encode rountrip.
---
--- Note: this also indirectly tests that 'deserialise' doesn't fail
--- on bytestrings produced by 'serialise'.
---
--- All (?) serialisation libraries guarantee this.
---
-decodeEncodeRoundtrip :: forall a.
-    (Eq a, Show a, Arbitrary a, SomeSerialisationConstraint a)
-    => Proxy a -> TestTree
-decodeEncodeRoundtrip _ = testProperty "decode-encode roundtrip" prop where
-    prop :: a -> Property
-    prop x = deserialise (serialise x) === x
-
--- | Strong 'deserialise' check.
--- We check that 'deserialise' is indeed total function.
---
--- Hardly any general serialisation library guarantee this.
---
-totalDeserialiseProp :: forall a.
-    (NFData a, SomeSerialisationConstraint a)
-    => Proxy a -> TestTree
-totalDeserialiseProp _ = testProperty "deserialise total" prop where
```

Comment:
We had this property. It should be easy to keep around, just make `Word64` deserialise to e.g. always `0` if there are too little data (instead of `error`).

Reply:
Is this generally a desirable property for all instances or do you just want to have it for some specific instances only, like `Word64`? Generally, I'm not too sold on silent failure, but it might be possible to convince me it's useful here.

Reply:
Recall, *we* don't use `deserialiseKey` anywhere in the internals. The only case I can remember is at the edge, when we (will) decode the actual results to the user (and in fact, I think only range lookup actually strictly needs to decode "in-between").

That would be actually an argument to not have `deserialiseKey` at all, and only require an extra class for range lookups.

But there it makes sense to be total, so the user gets something for their query, even if `deserialiseKey` fails (mistakes happen, it's user code, and can be incorrect).

Reply:
Ditto, `deserialiseValue`, mostly required at the edge only (but unfortunately also in monoidal tables).

And there if `deserialiseValue` behaves badly, whole dataset may be corrupted.

Reply:
> Recall, we don't use deserialiseKey anywhere in the internals. The only case I can remember is at the edge

> Ditto, deserialiseValue, mostly required at the edge only (but unfortunately also in monoidal tables).

As I understand it that is true for `SerialiseKey` and `SerialiseValue`, they only occur at the edge. We should be able to present a low level API without them, just plain `SerialisedKey/Value/Blob` (`WriteBuffer` is currently an exception, but there's already a TODO about removing the type parameters). It might make sense to expose that in a separate module and then layer the typed API on top. So `error` here should not cause any problems.

You are right that monoidal merging deserves special attention. Luckily we have a separate typeclass where we can isolate the unsafety, document and test additional requirements. The method in placeholder `SomeUpdateConstraint a` currently works on `a`, but I think the idea is that we want it to be defined on `SerialisedValue`. And then we can require this function to be total.

Of course it will be convenient for users to have a helper function `SerialiseValue v => (v -> v -> v) -> SerialisedValue -> SerialisedValue -> SerialisedValue`, but that again can have comments warning to only use it for `v` with total `deserialiseValue`.

Reply:
> That would be actually an argument to not have deserialiseKey at all, and only require an extra class for range lookups.

It's a nice observation that we could get away without `deserialiseKey` at all, except for range queries. For batch lookups we can remember the input keys and just match them up with the results.
I'm not sure it's worth the complexity and don't want to just change the class without further discussion, but at least good to keep in mind. I can leave a comment and bring it up with the others.

> But there it makes sense to be total, so the user gets something for their query, even if `deserialiseKey` fails (mistakes happen, it's user code, and can be incorrect).

I don't see many cases where the results are still useful when the keys are incorrect/placeholders. Carrying on with incorrect results might be worse than failing and users are not likely to handle the case properly (e.g. check for `key == 0`), unless it's clear from the types. We could make `deserialiseKey` return `Either` and use that to propagate errors nicely in some way?

Reply:
In any case, it's an interesting discussion and it could make sense to tweak the serialisation classes, but it's a much bigger topic than what I'm doing in this PR, which was mainly to prepare the macro benchmark. I'll merge now, but I'm happy to continue discussing here or elsewhere.